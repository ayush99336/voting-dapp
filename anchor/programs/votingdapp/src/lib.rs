#![allow(clippy::result_large_err)]

use anchor_lang::prelude::*;

declare_id!("6XeRbsLezt9HC8BzvczsTZFZ4CuZtT1j4ujwWhr34JMo");

#[program]
pub mod votingdapp {

    use super::*;
    pub fn initialize_poll(
        ctx: Context<InitializePoll>,
        poll_id: u64,
        description: String,
        poll_start: u64,
        poll_end: u64,
    ) -> Result<()> {
        let poll = &mut ctx.accounts.poll;
        poll.poll_id = poll_id;
        poll.description = description;
        poll.poll_start = poll_start;
        poll.poll_end = poll_end;

        Ok(())
    }

    pub fn initialize_candidate(
        ctx: Context<InitializeCandidate>,
        candidate_name: String,
        _poll_id: u64,
    ) -> Result<()> {
        let candidate = &mut ctx.accounts.candidate;
        candidate.candidate_name = candidate_name;
        ctx.accounts.poll.poll_option_index += 1;

        Ok(())
    }
    pub fn vote(ctx: Context<Vote>, _candidate_name: String, _poll_id: u64) -> Result<()> {
        let candidate_account = &mut ctx.accounts.candidate;

        let poll = &mut ctx.accounts.poll;
        let current_time = Clock::get()?.unix_timestamp;
        if current_time < poll.poll_start as i64 {
            return Err(ErrorCode::PollNotStarted.into());
        }
        if current_time > poll.poll_end as i64 {
            return Err(ErrorCode::PollEnded.into());
        }
        candidate_account.candidate_votes += 1;
        Ok(())
    }
}
#[derive(Accounts)]
#[instruction(poll_id:u64)]
pub struct InitializePoll<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,
    #[account(init,payer=signer,space=8+Poll::INIT_SPACE,seeds=[poll_id.to_le_bytes().as_ref()],bump)]
    pub poll: Account<'info, Poll>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(candidate_name:String,poll_id:u64)]
pub struct InitializeCandidate<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,
    #[account(seeds=[poll_id.to_le_bytes().as_ref()],bump)]
    pub poll: Account<'info, Poll>,
    #[account(init,payer=signer,space=8+Candidate::INIT_SPACE,seeds=[poll_id.to_le_bytes().as_ref(),candidate_name.as_ref()],bump)]
    pub candidate: Account<'info, Candidate>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(candidate_name:String,poll_id:u64)]
pub struct Vote<'info> {
    #[account(mut)]
    pub voter: Signer<'info>,
    // candidate PDA seed order must match InitializeCandidate (poll_id, candidate_name)
    #[account(mut, seeds=[poll_id.to_le_bytes().as_ref(), candidate_name.as_ref()],bump)]
    pub candidate: Account<'info, Candidate>,
    #[account(seeds=[poll_id.to_le_bytes().as_ref()],bump)]
    pub poll: Account<'info, Poll>,

    pub system_program: Program<'info, System>,
}

#[account]
#[derive(InitSpace)]
pub struct Candidate {
    #[max_len(32)]
    pub candidate_name: String,
    pub candidate_votes: u64,
}
#[account]
#[derive(InitSpace)]
pub struct Poll {
    pub poll_id: u64,
    #[max_len(280)]
    pub description: String,
    pub poll_start: u64,
    pub poll_end: u64,
    pub poll_option_index: u64,
}
#[error_code]
enum ErrorCode {
    PollNotStarted,
    PollEnded,
}
