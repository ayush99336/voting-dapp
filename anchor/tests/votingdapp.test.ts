import { Votingdapp } from '../target/types/votingdapp'
import { Program } from '@coral-xyz/anchor'
import { Keypair, PublicKey } from '@solana/web3.js'
import * as anchor from '@coral-xyz/anchor'
import { startAnchor } from 'solana-bankrun'
import { BankrunProvider } from 'anchor-bankrun'
const votingAddress = new PublicKey('JAVuBXeBZqXNtS73azhBDAoYaaAFfo4gWXoZe2e7Jf8H')
import IDL from '../target/idl/votingdapp.json'
describe('votingdapp', () => {
  let context
  let provider
  let votingProgram
  beforeAll(async () => {
    context = await startAnchor('', [{ name: 'votingdapp', programId: votingAddress }], [])
    provider = new BankrunProvider(context)
    votingProgram = new Program<Votingdapp>(IDL, provider)
  })

  it('Initialize Poll', async () => {
    const pollId = new anchor.BN(1)

    // derive the PDA for the poll using the poll_id seed (u64 little-endian)
    const pollSeed = Buffer.from(pollId.toArray('le', 8))
    const [pollPda] = PublicKey.findProgramAddressSync([pollSeed], votingAddress)

    await votingProgram.methods
      .initializePoll(pollId, 'What is your favorite peanut butter?', new anchor.BN(0), new anchor.BN(1856131203))
      .rpc()

    const [pollAddress] = PublicKey.findProgramAddressSync([pollSeed], votingAddress)
    const poll = await votingProgram.account.poll.fetch(pollAddress)
    console.log('Poll Account: ', poll)
    console.log(typeof poll.pollId)
    expect(poll.pollId.toNumber()).toEqual(1)
    expect(poll.description).toEqual('What is your favorite peanut butter?')
    expect(poll.pollStart.toNumber()).toBeLessThan(poll.pollEnd.toNumber())
  })
  it('Initialize Candidate', async () => {
    await votingProgram.methods.initializeCandidate('Smooth', new anchor.BN(1)).rpc()
    await votingProgram.methods.initializeCandidate('Crunchy', new anchor.BN(1)).rpc()

    const [smoothAddress] = PublicKey.findProgramAddressSync(
      [Buffer.from(new anchor.BN(1).toArray('le', 8)), Buffer.from('Smooth')],
      votingAddress,
    )
    const [crunchyAddress] = PublicKey.findProgramAddressSync(
      [Buffer.from(new anchor.BN(1).toArray('le', 8)), Buffer.from('Crunchy')],
      votingAddress,
    )

    const smoothCandidate = await votingProgram.account.candidate.fetch(smoothAddress)
    const crunchyCandidate = await votingProgram.account.candidate.fetch(crunchyAddress)
    console.log('Smooth Candidate: ', smoothCandidate)
    console.log('Crunchy Candidate: ', crunchyCandidate)
    expect(smoothCandidate.candidateName).toEqual('Smooth')
    expect(smoothCandidate.candidateVotes.toNumber()).toEqual(0)
    expect(crunchyCandidate.candidateName).toEqual('Crunchy')
    expect(crunchyCandidate.candidateVotes.toNumber()).toEqual(0)
  })

  it('Vote candidate', async () => {
    await votingProgram.methods.vote('Smooth', new anchor.BN(1)).rpc()
    const [smoothAddress] = PublicKey.findProgramAddressSync(
      [Buffer.from(new anchor.BN(1).toArray('le', 8)), Buffer.from('Smooth')],
      votingAddress,
    )
    const smoothCandidate = await votingProgram.account.candidate.fetch(smoothAddress)
    console.log('Smooth Candidate after vote: ', smoothCandidate)
    expect(smoothCandidate.candidateVotes.toNumber()).toEqual(1)
  })
})
