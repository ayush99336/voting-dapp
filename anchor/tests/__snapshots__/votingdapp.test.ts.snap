// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`votingdapp > should match the IDL 1`] = `
{
  "accounts": [
    {
      "discriminator": [
        110,
        234,
        167,
        188,
        231,
        136,
        153,
        111,
      ],
      "name": "Poll",
    },
  ],
  "address": "JAVuBXeBZqXNtS73azhBDAoYaaAFfo4gWXoZe2e7Jf8H",
  "instructions": [
    {
      "accounts": [
        {
          "name": "signer",
          "signer": true,
          "writable": true,
        },
        {
          "name": "poll",
          "pda": {
            "seeds": [
              {
                "kind": "arg",
                "path": "poll_id",
              },
            ],
          },
          "writable": true,
        },
        {
          "address": "11111111111111111111111111111111",
          "name": "system_program",
        },
      ],
      "args": [
        {
          "name": "poll_id",
          "type": "u64",
        },
        {
          "name": "description",
          "type": "string",
        },
        {
          "name": "poll_start",
          "type": "u64",
        },
        {
          "name": "poll_end",
          "type": "u64",
        },
        {
          "name": "canditate_amount",
          "type": "u64",
        },
      ],
      "discriminator": [
        193,
        22,
        99,
        197,
        18,
        33,
        115,
        117,
      ],
      "name": "initialize_poll",
    },
  ],
  "metadata": {
    "description": "Created with Anchor",
    "name": "votingdapp",
    "spec": "0.1.0",
    "version": "0.1.0",
  },
  "types": [
    {
      "name": "Poll",
      "type": {
        "fields": [
          {
            "name": "poll_id",
            "type": "u64",
          },
          {
            "name": "description",
            "type": "string",
          },
          {
            "name": "poll_start",
            "type": "u64",
          },
          {
            "name": "poll_end",
            "type": "u64",
          },
          {
            "name": "canditate_amount",
            "type": "u64",
          },
        ],
        "kind": "struct",
      },
    },
  ],
}
`;
